### PR TITLE
[staging] fetchurl: Support URLs with spaces in them

### PR DESCRIPTION
**Pull request by dasJ**
This is accomplished by switching to arrays instead of using
space-separated strings.

###### Motivation for this change

Download URLs with spaces from mirrors

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
- Please add a test case using [`invalidateFetcherByDrvHash`](https://nixos.org/manual/nixpkgs/unstable/#sec-pkgs-invalidateFetcherByDrvHash)
 - This could probably be simplified using structured attrs. Doesn't have to be scoped in afaic, but seems worth considering.

**Comment by roberth** [comment]
This doesn't have to go into staging, because it won't cause rebuilds.

**Comment by mohe2015** [comment]
How does this not cause rebuilds?

**Comment by dasJ** [comment]
`fetchurl` is usually used in a FOD context which is why nothing really changes.

@roberth I wanted to place this against staging so it's more likely for problems to get discovered without actually breaking master. Also I didn't want this to go into 21.11 because I would call `fetchurl` a release-critical package and those don't accept changes anymore iirc ;)
The test case is something I will do (although I'm kind of opposed to it because I'm too lazy ;)). The structured attrs will probably not work here because they generate JSON and I can already smell a cyclic dependency between `jq` and `fetchurl`.

File: pkgs/build-support/fetchurl/builder.sh [hunk 1]
```diff
@@ -120,18 +125,17 @@ for url in $urls; do
             if test -n "${!varName}"; then mirrors="${!varName}"; fi
 
             for url3 in $mirrors; do
-                urls2="$urls2 $url3$fileName";
+                urls2+=("$url3$fileName")
             done
         fi
     fi
 done
-urls="$urls2"
```
Comment (roberth) [inline pkgs/build-support/fetchurl/builder.sh:128]: Could you keep the assignment, or perhaps `set -u` and delete the old array?
Having to use `url2` seems like a mistake waiting to happen.

**Pull request closed without merging**